packages/contracts/src/contracts/current/protocol/Exchange/MixinMatchOrders.sol

```diff
@@ -0,0 +1,308 @@
+    /// @return matchedFillResults Amounts filled and fees paid by maker and taker of matched orders.
+    /// TODO: Make this function external once supported by Solidity (See Solidity Issues #3199, #1603)
+    function matchOrders(
+        Order memory leftOrder,
+        Order memory rightOrder,
+        bytes leftSignature,
+        bytes rightSignature
+    )
+        public
+        returns (MatchedFillResults memory matchedFillResults)
+    {
+        // Get left & right order info
+        OrderInfo memory leftOrderInfo = getOrderInfo(leftOrder);
+        OrderInfo memory rightOrderInfo = getOrderInfo(rightOrder);
+
+        // Fetch taker address
+        address takerAddress = getCurrentContextAddress();
+
+        // Either our context is valid or we revert
+        assertValidMatch(leftOrder, rightOrder);
+
+        // Compute proportional fill amounts
+        matchedFillResults = calculateMatchedFillResults(
+            leftOrder,
+            rightOrder,
+            leftOrderInfo.orderStatus,
+            rightOrderInfo.orderStatus,
+            leftOrderInfo.orderFilledAmount,
+            rightOrderInfo.orderFilledAmount
+        );
+
+        // Validate fill contexts
+        assertValidFill(
+            leftOrder,
+            leftOrderInfo.orderStatus,
+            leftOrderInfo.orderHash,
+            takerAddress,
+            leftOrderInfo.orderFilledAmount,
+            matchedFillResults.left.takerAssetFilledAmount,
+            leftSignature
+        );
+        assertValidFill(
+            rightOrder,
+            rightOrderInfo.orderStatus,
+            rightOrderInfo.orderHash,
+            takerAddress,
+            rightOrderInfo.orderFilledAmount,
+            matchedFillResults.right.takerAssetFilledAmount,
+            rightSignature
+        );
+
+        // Settle matched orders. Succeeds or throws.
+        settleMatchedOrders(
+            leftOrder,
+            rightOrder,
+            takerAddress,
+            matchedFillResults
+        );
+
+        // Update exchange state
+        updateFilledState(
+            leftOrder,
+            takerAddress,
+            leftOrderInfo.orderHash,
+            leftOrderInfo.orderFilledAmount,
+            matchedFillResults.left
+        );
+        updateFilledState(
+            rightOrder,
+            takerAddress,
+            rightOrderInfo.orderHash,
+            rightOrderInfo.orderFilledAmount,
+            matchedFillResults.right
+        );
+
+        return matchedFillResults;
+    }
+
+    /// @dev Validates context for matchOrders. Succeeds or throws.
+    /// @param leftOrder First order to match.
+    /// @param rightOrder Second order to match.
+    function assertValidMatch(
+        Order memory leftOrder,
+        Order memory rightOrder
+    )
+        internal
+    {
+        // The leftOrder maker asset must be the same as the rightOrder taker asset.
+        require(
```

Comment:
I still think we should optimize here by just copying `leftOrder.makerAssetData` to `rightOrder.takerAssetData`. This is one of the bigger optimizations we can make throughout our codebase. Willing to punt on it for MVP and add a TODO though.

Reply:
I've added a todo to look into optimizations, although generally I do like being explicit and even redundant in security-related code. I'd be more comfortable doing this if we could explicitly test revert scenarios. For example: we expect this to fail with a bad signature".